### PR TITLE
feat: Migrate to TypeScript

### LINKED ISSUE #99: Migrate to TypeScript
The repository is currently using Flow for static type checking. It should use TypeScript instead.

Requirements:
- Library is fully written in TS (no JS files apart from node modules).
- Build script compiles TS into JS.
- Compiled JS is side-effects free (required by tree-shaking).
- Compiled JS exposes Individual modules (required by tree-shaking).
- Compiled JS includes types.
- Compiled JS includes source maps.

As this is a major overhaul of the library, it should be marked as a breaking change and version should bump to 2.0.0

### PR DESCRIPTION
**Pull request by seb-cr**
This is a first iteration of Lambda Wrapper v2, which contains a number of breaking changes.

The key aims are:

- Migrate to a TypeScript-only codebase
- Reconsider how dependency injection should work so that dependency types can be inferred
- Fully document the new version
- Keep other changes to the package's API to a minimum

A good place to start reviewing this is probably [the README](https://github.com/comicrelief/lambda-wrapper/tree/migrate-to-typescript#readme), which provides examples of how the new wrapper works.

For a complete list of changes, please read the [v2 migration guide](https://github.com/comicrelief/lambda-wrapper/blob/migrate-to-typescript/docs/migration/v2.md). If you spot anything missing from there, please comment!

The tests are significantly affected, as you'd expect from changing how dependency injection works, and while going through them I took the opportunity to make improvements where I found them hard to understand. Currently test coverage is slightly below `master` (70.1% vs 75.7%).

In terms of what we'll do with this moving forward: v2 will first be published on the beta release channel. We can then test it out in one or two real-life applications to see how easy or difficult it is to work with, and continue to release improvements to the beta channel until we are satisfied that it's ready for wider use.

Jira: [ENG-1710], [ENG-1725]
Closes #99

[ENG-1710]: https://comicrelief.atlassian.net/browse/ENG-1710?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/core/DependencyInjection.ts [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  private isConstructing = true;
+
+  constructor(
+    readonly config: LambdaWrapperConfig,
+    readonly event: any,
+    readonly context: Context,
+  ) {
+    this.dependencies = Object.fromEntries(
+      Object.entries(config.dependencies)
+        .map(([, Constructor]) => [Constructor.name, new Constructor(this)]),
```
Comment (pauldolden) [inline src/core/DependencyInjection.ts:33]: I think I mentioned this one prior to your leave, and I can't remember the outcome of the conversation, so just flagging it here again as a reminder:

could we use `Object.values` instead?

File: src/models/ResponseModel.ts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+ */
+export const DEFAULT_MESSAGE = 'success';
+
+/**
+ * Our standard response model for HTTP endpoints.
+ */
+export default class ResponseModel {
+  body: any;
+
+  code: any;
```
Comment (pauldolden) [inline src/models/ResponseModel.ts:23]: Could we use `statusCode` instead, just for clarity?

I think I mentioned this one before, and again can't remember what was discussed.

**Review by pauldolden — COMMENTED** [review]
I've been giving this a more thorough read through this morning. The documentation and the migration guide are really helpful and well written, thanks for that.

As I've said previously I think this is a huge improvement on v1, even disregarding the obvious benefits of making this library type safe. I think the typing is largely fine, though if I recall correctly I believe this is something you said you'd like to iterate on and improve where possible, down the line.

The APIs are much more streamlined and offer more clarity than the previous method of DI. I think this syntax particularly:

```
export default lambdaWrapper.wrap(async (di) => {
  const request = di.get(RequestService);
  // ...
});
```

Is really well designed, and abstracts away some of the more clunky design decisions from v1.

The only real thought I had, in terms of design decisions, is the passing of a class directly to the `di.get` function. I can't put my finger on exactly why, but passing in a class as essentially an identifier, feels a little unintuitive.  I think maybe, it could be that this could potentially leave room for some confusion over what instance of a class is actually being used within the Lambda.

This is not a hill I would die on, but I think it was a point potentially worth raising and discussing.

Another thought I had, as this is a wholesale rewrite of this library, would it be worthwhile to consider migrating to the `AWS SDK v3` as a means of future proofing this library a little. I know there was some fairly considerable API changes so it may be the case that this is not plausible, but I thought it would be worth mentioning as we are unlikely to get many opportunities to implement this change down the line should Amazon start to phase out v2.

I'm not sure the extent of work this would create across other services, but as introducing a new version of Lambda Wrapper would constitute a breaking change any way, it might be worth considering doing this working in one fell swoop.

Again not something I'm calling out as a must have, but just a thought.

Your comments on test coverage are fair, but as this would initially be released into a beta version, I think that is something that could be improved fairly quickly if and when we encounter some limitations or bugs when we start testing this.

I think the pruning of more service specific logic (like the `MarketingPreference` model) is a good decision keeping this package lean and extensible is the best way of designing it. I do think it would be a good idea though, to look at some sort of companion service (or some similiar implementation), in order to keep frequently used services from being rewritten in a number of different services.

On the whole though, I think you've done a brilliant job with this and it will certainly help in laying the foundations for a more wide-spread use of TypeScript across our services. I noticed @corinja, self-requested a review on this so it would be great to hear input he has.

**Comment by seb-cr** [comment]
Thanks @pauldolden!

> I think the typing is largely fine, though if I recall correctly I believe this is something you said you'd like to iterate on and improve where possible, down the line.

I'm struggling to remember now, but I think I ended up using a lot of `any` in the services to avoid substantial changes. One of the next steps after this PR will be to review some of those service methods and replace or remove the less type-safe ones.

> The only real thought I had, in terms of design decisions, is the passing of a class directly to the `di.get` function. I can't put my finger on exactly why, but passing in a class as essentially an identifier, feels a little unintuitive. I think maybe, it could be that this could potentially leave room for some confusion over what instance of a class is actually being used within the Lambda.

Understandable. The class constructor approach did feel weird at first but it was the winner in terms of simplicity and I've grown fond of it since. Rather than an identifier, which is how v1 operates, consider it like this: `di.get(MyClass)` gets me an instance of `MyClass`. There's not really a question of _which_ instance; all these service classes should be singletons. I'll check that I made this clear in the docs.

> Another thought I had, as this is a wholesale rewrite of this library, would it be worthwhile to consider migrating to the `AWS SDK v3` as a means of future proofing this library a little? I know there was some fairly considerable API changes so it may be the case that this is not plausible, but I thought it would be worth mentioning as we are unlikely to get many opportunities to implement this change down the line should Amazon start to phase out v2.
> 
> I'm not sure the extent of work this would create across other services, but as introducing a new version of Lambda Wrapper would constitute a breaking change any way, it might be worth considering doing this working in one fell swoop.

When updating other projects that _use_ Lambda Wrapper, it might indeed make sense in terms of maintenance effort to do both at once.

Within Lambda Wrapper, I believe our use of AWS SDK is all "under the hood" so updating it should be only a patch release, which can be done anytime. It's therefore not something I'd look at before the 2.0.0 release (unless AWS deprecate it sooner than expected).

> I do think it would be a good idea though, to look at some sort of companion service (or some other implementation), in order to keep frequently used services from being rewritten in a number of different repos, with slight variances in terms of implementation.

Agreed, and as you know I've been wanting to do this for a long time! Let's discuss separately in due course – could be a nice hackday project after RND.

File: src/core/DependencyInjection.ts (hunk 1 above)
Comment (seb-cr) [reply src/core/DependencyInjection.ts:33]: Yes, good spot. This is left over from an earlier iteration where I was using the keys for something, but now they are ignored.

File: src/models/ResponseModel.ts (hunk 2 above)
Comment (seb-cr) [reply src/models/ResponseModel.ts:23]: I agree, but am saving this for a future PR – trying to keep this one focussed on TypeScript and redesigning DI with as few other changes as possible!

File: src/core/DependencyInjection.ts (hunk 1 above)
Comment (seb-cr) [reply src/core/DependencyInjection.ts]: https://github.com/comicrelief/lambda-wrapper/pull/882/commits/a3d12b381cedd77ed879e6b044c1f3afd35d7314

**Comment by pauldolden** [comment]
All completely fair. As I mentioned on the call the other day, I didn't really have an awful lot to add on how you've gone about redesigning this. Those comments were largely points I thought worth discussing rather than a vehement desire to get changes made.

Like I said, I think this is all really good work and a big improvement on v1. Nice work! 🙂.

**Review by corinja — APPROVED** [review]
This is a great _and_ substantial piece of work. I've read the `readme` files and skimmed the code and I can't suggest any changes, nor can I spot anything that needs to be added. I think it'll be good to start migrating services to use this Lambda Wrapper in order to spot any flaws, but I don't think there are any significant ones. Very impressive, overall.

**Comment by seb-cr** [comment]
After a hiccup with CI and my getting the commit message wrong for this PR, this is now released as version 2.0.0-beta.1 :tada:

The release is available on:
- [npm package (@beta dist-tag)](https://www.npmjs.com/package/@comicrelief/lambda-wrapper/v/2.0.0-beta.1)
- [GitHub release](https://github.com/comicrelief/lambda-wrapper/releases/tag/v2.0.0-beta.1)

**Pull request merged**